win32ss/drivers/miniport/pc98vid/pc98vid.c

```diff
@@ -193,6 +193,15 @@ Pc98VidGetVideoChildDescriptor(
     return ERROR_NO_MORE_DEVICES;
 }
 
+#if defined(_MSC_VER)
+/*
+ * Avoid C2983 error for MSVC 2015. There is no such thing
+ * as DRIVER_INITIALIZE for video miniport drivers.
+ */
+#pragma alloc_text(INIT, DriverEntry)
+#else
+CODE_SEG("INIT")
+#endif
```

Comment:
I don't see how the INIT segment thing is related to DRIVER_INITIALIZE .

Reply:
Yes, Just put the CODE_SEG declaration

Reply:
Maybe DRIVER_INITIALIZE is a special linker thing, I had this error:
<details>
D:\a\reactos\reactos\src\win32ss\drivers\miniport\pc98vid\pc98vid.c(210): error C2983: 'DriverEntry': all declarations must have an identical __declspec(code_seg(...))
D:\a\reactos\reactos\src\sdk\include\ddk\video.h(62): note: see declaration of 'DriverEntry'
  
</details>

This declaration doesn't help
```c
CODE_SEG("INIT")
ULONG
NTAPI
DriverEntry(
    _In_ PVOID Context1,
    _In_ PVOID Context2);
```

Reply:
You need to sync
win32ss/drivers/miniport/pc98vid/pc98vid.c

Reply:
> I had this error:
> D:\a\reactos\reactos\src\win32ss\drivers\miniport\pc98vid\pc98vid.c(210): error C2983: 'DriverEntry': all declarations must have an identical __declspec(code_seg(...)) D:\a\reactos\reactos\src\sdk\include\ddk\video.h(62): note: see declaration of 'DriverEntry'

Because in our video.h it is declared like that:
https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/include/ddk/video.h;h=1500e03bb9bfac8827e03e2e6a9ab9482df6c822;hb=HEAD#l59
```c
  60 ULONG
  61 NTAPI
  62 DriverEntry(
  63   PVOID Context1,
  64   PVOID Context2);
  65 
```
without any CODE_SEG thing, which is actually legit because this is in the public DDK and it cannot enforce you to put the thing in a INIT section, but you want to, and with MSVC 2015 when using that CODE_SEG thing it would appear it requires as well the prototype to have the same CODE_SEG.

Reply:
We shouldn't add any CODE_SEG to the video.h's DriverEntry prototype, because otherwise we would break compatibility with the official MS DDK/WDK.

Reply:
Yes, that's why I added ` #if defined(_MSC_VER) `. I really don't understand why WDM stuff works there without CODE_SEG thing https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/include/xdk/iotypes.h;hb=eba00d1b5ed678c5557271649d7cb41dd4ee47f7#l2229

Reply:
It looks like the other miniports do not have this.
Need to update them all or _none_, due to a common declaration?
https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/include/ddk/video.h;hb=HEAD#l59

Reply:
> Yes, that's why I added `#if defined(_MSC_VER)`. I really don't understand why WDM stuff works there without CODE_SEG thing.

Because in that header for example, this is a (function) typedef that is being declared, and not an actual function prototype.

Reply:
Sure :sweat_smile:

Reply:
Then, forget about it or add it for 2017+ only.

Reply:
Maybe we should just have VS2015 have CODE_SEG being a NOP. We already spent way too much time on this...

Reply:
What do we need MSVC 2015 for? The bug was unnoticed for quite some time, nobody cared.
Now we have to change our practice and include CODE_SEG even in headers, just for this particular compiler

Reply:
The `#pragma alloc_text(...)` technique Microsoft always used (and still use), e.g. in the WDK, etc.., always worked, from MSVC 6 up to now.

Reply:
Yeah, but that doesn't work for GCC

Reply:
> The bug was unnoticed for quite some time

Not fully true: Jira tickets were/are filed about VS2015 regressions.
Of course, dropping VS2015 CI, while claiming support, did not help!

> nobody cared.

That, on the other hand...